### PR TITLE
GH-43187: [C++] Support basic is_in predicate simplification

### PR DESCRIPTION
**Pull request by larry98**
### Rationale for this change

Prior to https://github.com/apache/arrow/pull/43256, this PR adds a basic implementation that does a linear scan filter over the value set on each guarantee. This isolates the correctness/semantics of `is_in` predicate simplification from the binary search performance optimization.

### What changes are included in this PR?

`SimplifyWithGuarantee` now handles `is_in` expressions.

### Are these changes tested?

A new unit test was added to arrow-compute-expression-test testing this change.

### Are there any user-facing changes?

No.
* GitHub Issue: #43187

### CODE REVIEW

File: cpp/src/arrow/compute/expression_test.cc [hunk 1]
```diff
@@ -1616,6 +1616,82 @@ TEST(Expression, SimplifyWithComparisonAndNullableCaveat) {
...
+
+    Simplify{is_in(field_ref("i32"), int32(), "[1,3,5,7,9]", null_matching_behavior)}
+        .WithGuarantee(greater(field_ref("i32"), literal(0)))
+        .ExpectUnchanged();
+
+    Simplify{is_in(field_ref("i32"), int32(), "[1,3,5,7,9]", null_matching_behavior)}
+        .WithGuarantee(
+            or_(equal(field_ref("i32"), literal(3)), is_null(field_ref("i32"))))
+        .ExpectUnchanged();
+
```
Comment (bkietz) [inline cpp/src/arrow/compute/expression_test.cc:1672]: I'd expect this to pass:
```suggestion
    Simplify{is_in(field_ref("i32"), int32(), "[1,3,null]", SetLookupOptions::MATCH)}
        .WithGuarantee(
            or_(equal(field_ref("i32"), literal(3)), is_null(field_ref("i32"))))
        .Expect(is_in(field_ref("i32"), int32(), "[3,null]", SetLookupOptions::MATCH));
```

File: cpp/src/arrow/compute/expression_test.cc [hunk 2]
```diff
@@ -1616,6 +1616,82 @@ TEST(Expression, SimplifyWithComparisonAndNullableCaveat) {
...
+  }
+
+  Simplify{
+      is_in(field_ref("i32"), int32(), "[1,3,5,7,9]", SetLookupOptions::INCONCLUSIVE)
+  }
+      .WithGuarantee(greater(field_ref("i32"), literal(3)))
+      .ExpectUnchanged();
+}
+
 TEST(Expression, SimplifyThenExecute) {
```
Comment (bkietz) [inline cpp/src/arrow/compute/expression_test.cc:1695]: I'd like to see more tests with null in the value_set, and some tests running against actual arrays. `TEST(Expression, SimplifyThenExecute)` should serve as a guide

**Review by bkietz — CHANGES_REQUESTED** [review]
This is looking great, thanks!

File: cpp/src/arrow/compute/expression_test.cc (hunk 1 above)
Comment (larry98) [reply cpp/src/arrow/compute/expression_test.cc:1673]: The original intention was to not support simplification if nulls in the value set cannot be dropped (either the guarantee is nullable or the null matching behavior is `INCONCLUSIVE`). This is because in the optimized implementation where we binary search and slice the value set array, slicing the front would drop nulls (assuming they are placed at the end) so we would have to reallocate a new array for the simplified value set.

Do you think we ought to support nulls in the value set, and if so any thoughts on how we'd continue to support this with the binary search/slice implementation?

File: cpp/src/arrow/compute/expression_test.cc (hunk 2 above)
Comment (larry98) [reply cpp/src/arrow/compute/expression_test.cc:1695]: Added a test to actually execute against arrays. For value sets containing nulls, I left a comment about this above.

File: cpp/src/arrow/compute/expression_test.cc (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/compute/expression_test.cc:1673]: The filtering approach should be able to support arbitrary value sets, so it could serve as a fallback for the binary search/slice implementation

File: cpp/src/arrow/compute/expression_test.cc [hunk 3]
```diff
@@ -1643,6 +1719,44 @@ TEST(Expression, SimplifyThenExecute) {
...
+    Datum evaluated, simplified_evaluated;
+    ExpectExecute(filter, input, &evaluated);
+    ExpectExecute(simplified, input, &simplified_evaluated);
+    if (simplified_evaluated.is_scalar()) {
+      ASSERT_EQ(evaluated.kind(), Datum::ARRAY);
+      ASSERT_EQ(simplified_evaluated.type()->id(), Type::BOOL);
+      BooleanBuilder builder;
+      ASSERT_OK(builder.AppendValues(
+          evaluated.length(), simplified_evaluated.scalar_as<BooleanScalar>().value));
+      ASSERT_OK_AND_ASSIGN(simplified_evaluated, builder.Finish());
```
Comment (bkietz) [inline cpp/src/arrow/compute/expression_test.cc:1754]:
```suggestion
      ASSERT_OK_AND_ASSIGN(simplified_evaluated,
          MakeArrayFromScalar(*simplified_evaluated.scalar(), expected->length()));
```

File: cpp/src/arrow/compute/expression_test.cc [hunk 4]
```diff
@@ -1643,6 +1719,44 @@ TEST(Expression, SimplifyThenExecute) {
   AssertDatumsEqual(evaluated, simplified_evaluated, /*verbose=*/true);
 }
 
+TEST(Expression, SimplifyIsInThenExecute) {
```
Comment (bkietz) [inline cpp/src/arrow/compute/expression_test.cc:1722]: Thank you for adding this!

**Review by bkietz — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/compute/expression_test.cc (hunk 1 above)
Comment (larry98) [reply cpp/src/arrow/compute/expression_test.cc:1662]: I see, that sounds reasonable to me. I added new tests for nullable guarantees and nulls in the value set for all of the different null matching behaviors.

**Comment by larry98** [comment]
@bkietz any ideas on the failing checks? The errors don't appear to be relevant, but lmk if I should look into them.

**Comment by mapleFU** [comment]
Failed CI unrelated

**Review by bkietz — APPROVED** [review]
LGTM, thanks for adding this!
